### PR TITLE
[AutoDiff] Create JVPEmitter

### PR DESCRIPTION
**Pull request by bartchr808**
Creates a very barebones `JVPEmitter` which copies the body of the original function returning `(original, undef)`, and creates the differential function with a body that returns an `undef`.

Apologies for the weird commit history, there was something wrong with my `tensorflow` branch on the fork of this repo that caused this. However, I fixed it now so future PRs won't be this ugly.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -5875,108 +6123,104 @@ ADContext::declareExternalAssociatedFunction(
   return assocFn;
 }
 
-static SILFunction* createJVP(
+static SILFunction *createEmptyVJP(
```
Comment (bartchr808) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Renamed this to more closely resemble naming and logic of `VJPEmitter` which creates an empty VJP and then uses an emitter to generate the body

Comment (bartchr808) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: The git diff here is acting funny in the next several lines, may be hard to follow but I more so made a copy of `createEmptyVJP` and named it `createEmptyJVP` and removed the previous `createJVP` to more closely resemble how `VJPEmitter` looks like and `JVPEmitter` will become.

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
...
+
+  /// The differential function.
+  SILFunction *differential;
+
+  /// The differentiation invoker.
+  DifferentiationInvoker invoker;
+
+  /// Caches basic blocks whose phi arguments have been remapped (adding a
+  /// predecessor enum argument).
+  SmallPtrSet<SILBasicBlock *, 4> remappedBasicBlocks;
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: This data structure is unnecessary.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
...
+
+  /// Temporary boolean for determine whether we should be generating the body of the JVP.
+  /// If true, then skip the JVP generation process since we cannot generate JVPs for the
+  /// primitive VJPs defined on addition for Float for example.
+  bool vjpGenerated;
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: It would be nicer if this is checked by the caller of `JVPEmitter` instead of being leaked into `JVPEmitter`

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 4]
```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
...
+
+    SmallVector<SILParameterInfo, 8> diffParams;
+    SmallVector<SILResultInfo, 8> diffResults;
+    auto origParams = origTy->getParameters();
+    auto indices = attr->getIndices();
+
+    // Add differential result for the seed.
+    auto origResInfo = origTy->getResults()[indices.source];
+    diffResults.push_back(SILResultInfo(
+                              origResInfo.getType()
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I feel the indentation is very awkward here. The convention we follow in the AD code base is usually indenting each line break by exactly 4 regardless of the parent expression's start position.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 5]
```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
...
+                                  ->getAutoDiffAssociatedTangentSpace(
+                                        lookupConformance)
+                                  ->getCanonicalType(),
+                              origResInfo.getConvention()));
+
+    // Add pullback results for the requested wrt parameters.
+    for (auto i : indices.parameters->getIndices()) {
+      auto origParam = origParams[i];
+      diffParams.push_back(SILParameterInfo(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I'd probably format it this way:

```cpp
      diffParams.push_back(
          SILParameterInfo(origParam.getType()
              ->getAutoDiffAssociatedTangentSpace(lookupConformance)
	      ->getCanonicalType(), origParam.getConvention()));
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 6]
```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
...
+                                    SILDebugScope(original->getLocation(),
+                                                  differential));
+    // Create empty body of differential.
+    auto diffConv = differential->getConventions();
+    auto *entry = differential->createBasicBlock();
+    createEntryArguments(differential);
+    // Return undef.
+    SILBuilder builder(entry);
+    auto loc = differential->getLocation();
+    builder.createReturn(loc, SILUndef::get(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I'd format it this way to for consistency with the rest of the code base.
```cpp
builder.createReturn(loc, SILUndef::get(
    differential->mapTypeIntoContext(diffConv.getSILResultType()),
    *differential));
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 7]
```diff
@@ -3427,6 +3428,217 @@ class VJPEmitter final
...
+  bool run();
+
+  void postProcess(SILInstruction *orig, SILInstruction *cloned) {
+    if (errorOccurred)
+      return;
+    SILClonerWithScopes::postProcess(orig, cloned);
+  }
+
+  /// Remap original basic blocks, adding predecessor enum arguments.
+  SILBasicBlock *remapBasicBlock(SILBasicBlock *bb) {
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: No need for this method. We can add it back when we do need it.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 8]
```diff
@@ -5848,6 +6060,41 @@ bool VJPEmitter::run() {
...
+    if (errorOccurred)
+      return true;
+  } else {
+    // Create empty body of JVP if the user defined their own custom VJP.
+    // Return undef.
+    auto diffConv = jvp->getConventions();
+    SILBuilder builder(entry);
+    auto loc = jvp->getLocation();
+    builder.createReturn(loc, SILUndef::get(
+                                  jvp->mapTypeIntoContext(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Fix indentation.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 4 above)
Comment (bartchr808) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Fair, I will put
```swift
diffResults.push_back(SILResultInfo(origResInfo.getType()
        ->getAutoDiffAssociatedTangentSpace(lookupConformance)
        ->getCanonicalType(), origResInfo.getConvention()));
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 3 above)
Comment (bartchr808) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I saw `bool JVPEmitter::run()` as the main function that creates a function, which in this case is called from `processDifferentiableAttribute`. Are you suggesting to add two different `run()` functions, one for if the VJP was generated and another if it wasn't? So I can then call either one of them in `processDifferentiableAttribute`?

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Sorry, I meant to suggest not creating a `JVPEmitter` at all for JVPs that are supposed to return `undef`.

Comment (bartchr808) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Ahh I see sounds good!

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 9]
```diff
@@ -3427,6 +3428,198 @@ class VJPEmitter final
...
+    SILOptFunctionBuilder fb(context.getTransform());
+    // The generated tangent linkage is set to Hidden because generated tangent
+    // are never called cross-module.
+    auto linkage = SILLinkage::Hidden;
+    auto *differential = fb.createFunction(
+        linkage, diffName, diffType, diffGenericEnv, original->getLocation(),
+        original->isBare(), IsNotTransparent, original->isSerialized(),
+        original->isDynamicallyReplaceable());
+    differential->setOwnershipEliminated();
+    differential->setDebugScope(new (module)
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```cpp
differential->setDebugScope(
    new (module) SILDebugScope(original->getLocation(), differential));
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 10]
```diff
@@ -3427,6 +3428,198 @@ class VJPEmitter final
...
+    // Return a tuple of the original result and an undef, which at some point
+    // will be the differential.
+    SmallVector<SILValue, 8> directResults;
+    directResults.append(origResults.begin(), origResults.end());
+
+    // Get differential result type.
+    auto jvpResultArray = jvp->getLoweredFunctionType()->getResults();
+    auto funcType = jvpResultArray.back().getType();
+    auto silFuncCanType = funcType->castTo<SILFunctionType>()
+                              ->getCanonicalType();
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
        ->getCanonicalType();
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 11]
```diff
@@ -3427,6 +3428,198 @@ class VJPEmitter final
...
+
+    directResults.push_back(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```cpp
directResults.push_back(
    SILUndef::get(jvp->mapTypeIntoContext(
        SILType::getPrimitiveObjectType(silFuncCanType)), *jvp));
```

**Review by rxwei — APPROVED** [review]

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow
